book/multi-arch-controller.md

```diff
@@ -0,0 +1,114 @@
+# Multi Arch Controller
+
+
+## Overview
+
+The Multi Arch Controller is a Kubernetes Controller that can provision virtual machines for Tekton `TaskRun` instances. This allows these TaskRuns to SSH into the provided virtual machines to build images in architectures other than amd64.
+
+## The TaskRun Contract
+
+The Multi Arch Controller has no understanding of the build pipeline logic. It only understands how to provision hosts and allocate SSH keys. For the Multi Arch Controller to act on a `TaskRun` it must meet the following conditions:
+
+- It must have the `build.appstudio.redhat.com/multi-arch-required` label
+- It must mount a secret with the name `multi-arch-ssh-$(context.taskRun.name)`
+- It must have a parameter called `ARCH` which specifies the required architecture
+
+If these conditions are met the controller will reconcile on the task and attempt to provision a virtual machine for the task. If it is successful it will create  secret (`multi-arch-ssh-$(context.taskRun.name)`) with the following entries:
+
+- `id_rsa` the SSH key to use to connect to the host
+- `host` the host name or IP to connect to, including the username in the form `user@host`
+- `build-dir` the working directory to perform the build in
+
+Note that these SSH credentials are only valid while the `TaskRun` is executing.
+
+If the operator is unsuccessful then it will still create the secret, however it will only have a single item called `error` with the error message. The task is expected to echo this to the logs and then fail if this error is present.
+
+The reason why this works is that the `TaskRun` will create a pod that references a non-existent secret (e.g. if the `TaskRun` was called `devfile-sample-t7rd7-build-container-arm64` then the resulting pod will reference a secret called `multi-arch-ssh-devfile-sample-t7rd7-build-container-arm64`). As this secret does not exist yet, the pod will not start but instead Kubernetes will wait until the secret is ready before running the pod. This gives the controller time to provision a SSH based host, and then create the secret with the required details. The secret is created even on failure to prevent the task hanging until timeout and to allow any error messages to be propagated to the user.
```

Comment:
Is there a scheduler timeout that may kick in prematurely?

Reply:
There does not seem to be. The timeout that kicks in is the TaskRun timeout.